cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -77,6 +82,66 @@ def merge_data(local_data, remote_data)
       merged
     end
 
+    def dry_report(merged)
+      added_services = []
+      removed_services = []
+      kept_services = {}
+      merged.each do |stackname, data|
+        if data[:remote].nil?
+          dry_msg "New stack #{stackname} would be installed to master"
+          added_services = data[:local][:stack]['services'].map { |svc| "#{stackname}/#{svc['name']}" }
+        elsif data[:local]
+          data[:local][:stack]['services'].each do |svc|
+            remote_svc = data[:remote]['services'].find { |s| s['name'] == svc['name'] }
+            if remote_svc
+              kept_services["#{stackname}/#{svc['name']}"] ||= {}
+              #kept_services["#{stackname}/#{svc['name']}"][:before] = JSON.load(JSON.dump(remote_svc)) # naughty deep clone/dup
+              #kept_services["#{stackname}/#{svc['name']}"][:after] = JSON.load(JSON.dump(svc))
+            else
+              added_services << "#{stackname}/#{svc['name']}"
+            end
+          end
+        end
+      end
+
+      merged.each do |stackname, data|
+        if data[:local].nil?
+          dry_msg "Stack  #{stackname} would be removed from master"
+          removed_services = data[:remote]['services'].map { |svc| "#{stackname}/#{svc['name']}" }
+        else
+          data[:local][:stack]['services'].each do |svc|
+            unless data[:remote]['services'].find { |s| s['name'] == svc['name'] }
+              added_services << "#{stackname}/#{svc['name']}"
+            end
+          end
+          data[:remote]['services'].each do |svc|
+            unless data[:local][:stack]['services'].find { |s| s['name'] == svc['name'] }
+              removed_services << "#{stackname}/#{svc['name']}"
+            end
+          end
+        end
+      end
+
+      unless added_services.empty?
+        dry_msg "The following services would be added to master"
+        added_services.each { |svc| puts " - #{svc}" }
+      end
+
+      unless removed_services.empty?
+        dry_msg "The following services would be removed from master"
+        removed_services.each { |svc| puts " - #{svc}" }
+      end
+
+      unless kept_services.empty?
+        dry_msg "The following services would be kept:"
```

Comment:
Better to say `The following services would be upgraded`, because the services are not just kept, they can change